File: Problem_2.js
// Basic Calculator || (https://leetcode.com/problems/basic-calculator-ii/)

// Without Stack
// Time Complexity : O(N)
// Space Complexity : O(1)

// With Stack
// Time Complexity : O(N)
// Space Complexity : O(N)

// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

/**
 * @param {string} s
 * @return {number}
 */
var calculate = function (s) {
    if (s === null || s.length === 0)
        return 0;

    // W/o stack

    // let tail = 0; 
    // let lastSign = "+";
    // let num = 0;
    // let result = 0;
    // for(let i=0; i<s.length; i++){
    //     let ch = s[i];
    //     if(!isNaN(ch) && ch!==" " ){
    //         num = num*10 + parseInt(ch);
    //     }
    //     if((isNaN(ch) && ch!==" ") || (i === s.length-1)){
    //         switch(lastSign){
    //             case "+": result = result + num;
    //                 tail = num;
    //                 break;
    //             case "-": result = result - num;
    //                 tail = -num;                
    //                 break;
    //             case "*": result = result - tail + (tail*num);
    //                 tail = tail*num;
    //                 break;
    //             case "/": 
    //                 let newTail = tail<0 ? Math.ceil(tail/num) : Math.floor(tail/num);
    //                 result = (result - tail) + newTail;
    //                 tail = newTail;
    //                 break;
    //         }
    //         num = 0;
    //         lastSign = ch;
    //     }
    // }
    // result+=num;
    // return result;

    // With stack
    let stack = [];
    let result = 0;
    s = s.trim();
    let num = 0;
    let lastSign = "+"
    for (let i = 0; i < s.length; i++) {
        let ch = s[i];
        // Grab number
        if (!isNaN(ch) && ch !== " ") {
            num = num * 10 + parseInt(ch);
        }
        // If found a sign, push to stack
        // If found / or *, pop from stack, calculate and then push
        if ((isNaN(ch) && ch !== " ") || (i === s.length - 1)) {
            switch (lastSign) {
                case "+": stack.push(num)
                    break;
                case "-": stack.push(-num)
                    break;
                case "*": stack.push(stack.pop() * num);
                    break;
                case "/":
                    let lastNum = stack.pop();
                    let newNum = lastNum < 0 ? Math.ceil(lastNum / num) : Math.floor(lastNum / num);
                    stack.push(newNum);
                    break;
            }
            num = 0;
            lastSign = ch;

        }
    }
    // Add all elements in stack
    while (stack.length > 0) {
        result += stack.pop();
    }
    return result;
};

File: problem2.cpp
/*
// Time Complexity : O(N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach

if bracket found solve the inside and give it as a number 
find the number and previous operator.
as soon as the next operator is found store the number and the last operator in stack.
once completed all the string
add all the elements stored in the stack

*/

#include<iostream>
#include<vector>
#include<string>
#include<stack>

using namespace std;

// self solution

class Solution {
    vector<int> helper(string& s,int st_idx){
        vector<int> res(2,0);
        stack<int> st;
        int curr_num{0};
        char last_expr{'+'};
        //cout<<"curr_num "<<"last_expr"<<endl;
        int i{st_idx};
        for(;s.at(i)!=')';++i){
            //cout<<curr_num<<" "<<last_expr<<endl;
            char c = s.at(i);
            if((isdigit(c)) || isspace(c) || c == '('){
                if(isdigit(c)) curr_num = curr_num * 10 + (c - '0');
                else if(c == '('){
                    vector<int> arr{0,0};
                    arr = helper(s,i+1);
                    curr_num = arr.at(0);
                    i = arr.at(1);
                }
            }
            else if(last_expr == '+' || last_expr == '-'){
                //cout<<"curr_num "<<curr_num<<endl;
                int mul = (last_expr == '+')?1:-1;
                st.push(mul*curr_num);
                last_expr = s.at(i);
                curr_num = 0; 
            }
        }
        int mul = (last_expr == '+')?1:-1;
        st.push(mul*curr_num);
        int sum{};
        //cout<<"element inside the stacks "<<endl;
        while(!st.empty()){
            int num = st.top();
            //cout<<num<<" ";
            sum+= num;
            st.pop();
        }
        cout<<endl;
        res.at(0) = sum;
        res.at(1) = i;
        return res;
    }
public:
    int calculate(string s) {
        int curr_num{0};
        char last_expr{'+'};
        int sz = s.size();
        stack<int> st{};
        for(int i{};i<sz;++i){
            char c = s.at(i);
            
            if((isdigit(c)) || isspace(c) || c == '('){
                if(isdigit(c)) curr_num = curr_num * 10 + (c - '0');
                if(c == '('){
                    //cout<<"start_bracket "<<i<<endl;;
                    vector<int> arr{0,0};
                    arr = helper(s,i+1);
                    curr_num = arr.at(0);
                    i = arr.at(1);
                    //cout<<"close_bracket "<<i<<endl;
                    //cout<<"curr_num "<<curr_num<<endl;
                }
            }
            else if(last_expr == '+' || last_expr == '-'){
                int mul = (last_expr == '+')?1:-1;
                st.push(mul*curr_num);
                last_expr = s.at(i);
                curr_num = 0; 
            }
        }
        int mul = (last_expr == '+')?1:-1;
        st.push(mul*curr_num);
        int sum{};
        while(!st.empty()){
            int num = st.top();
            sum+=num;
            st.pop();
        }
        return sum;

    }
};




